File: ios/Objective-C/GeneralSettingsObjC/GeneralSettingsObjC/Modules/PullListView/PullListView.m
//
//  PullListView.m
//  GeneralSettings
//
//  Created by dynamsoft on 2021/11/29.
//

#import "PullListView.h"

@interface PullListView ()<UIPickerViewDelegate, UIPickerViewDataSource>
{
    NSString *pullViewTitle;
    NSArray *pullListDataArray;
    
    /// Record currentSelectedDic.
    NSDictionary *recordCurrerntSelectedDic;
}

@property (nonatomic, strong) UIView *backgroundMaskView;

@property (nonatomic, strong) UIView *pullListBackgroundView;

@property (nonatomic, strong) UIView *headerView;

@property (nonatomic, strong) UIButton *cancelButton;

@property (nonatomic, strong) UIButton *confirmButton;

@property (nonatomic, strong) UILabel *titleLabel;

@property (nonatomic, strong) UIPickerView *pickerView;

/// Selected completion block.
@property (nonatomic, copy) void(^completion)(NSDictionary *selectedDic);

@end

@implementation PullListView

+ (void)showPullListViewWithArray:(NSArray *)pullListArray titleName:(NSString *)title selectedDicInfo:(NSDictionary *)selectedDicInfo completion:(void (^)(NSDictionary *selectedDic))completion
{
    PullListView *pullListView = [[PullListView alloc] initWithFrame:CGRectMake(0, 0, kScreenWidth, kScreenHeight) WithArray:pullListArray titleName:title selectedDicInfo:selectedDicInfo completion:completion];
    
    [pullListView show];
}

- (instancetype)initWithFrame:(CGRect)frame WithArray:(NSArray *)pullListArray titleName:(NSString *)title selectedDicInfo:(NSDictionary *)selectedDicInfo completion:(void (^)(NSDictionary *selectedDic))completion
{
    self = [super initWithFrame:frame];
    if (self) {
        pullViewTitle = title;
        pullListDataArray = pullListArray;
        recordCurrerntSelectedDic = selectedDicInfo;
        self.completion = completion;
        [self setupUI];
    }
    return self;
}

- (void)setupUI
{
    self.backgroundMaskView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, kScreenWidth, self.height)];
    self.backgroundMaskView.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.2];
    self.backgroundMaskView.userInteractionEnabled = YES;
    [self addSubview:self.backgroundMaskView];
    
    CGFloat pullListHeight = 0;
    if (kIs_iPhoneXAndLater) {
        pullListHeight = 270 * kScreenAdaptationRadio + 34;
    } else {
        pullListHeight = 270 * kScreenAdaptationRadio;
    }
    
    self.pullListBackgroundView = [[UIView alloc] initWithFrame:CGRectMake(0, kScreenHeight - pullListHeight, kScreenWidth, pullListHeight)];
    self.pullListBackgroundView.backgroundColor = [UIColor whiteColor];
    [self.backgroundMaskView addSubview:self.pullListBackgroundView];
    
    self.headerView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, kScreenWidth, 40 * kScreenAdaptationRadio)];
    self.headerView.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent:0.05];
    [self.pullListBackgroundView addSubview:self.headerView];
    
    self.titleLabel = [[UILabel alloc] initWithFrame:CGRectMake((kScreenWidth - 100 * kScreenAdaptationRadio) / 2.0, 0, 100 * kScreenAdaptationRadio, self.headerView.height)];
    self.titleLabel.text = pullViewTitle;
    self.titleLabel.textColor = [UIColor grayColor];
    self.titleLabel.font = kFont_Regular(13 * kScreenAdaptationRadio);
    self.titleLabel.textAlignment = NSTextAlignmentCenter;
    [self.headerView addSubview:self.titleLabel];
    
    self.cancelButton = [UIButton buttonWithType:UIButtonTypeCustom];
    self.cancelButton.frame = CGRectMake(0, 0, 60 * kScreenAdaptationRadio, self.headerView.height);
    [self.cancelButton setTitle:@"Cancel" forState:UIControlStateNormal];
    [self.cancelButton setTitleColor:[UIColor grayColor] forState:UIControlStateNormal];
    self.cancelButton.titleLabel.font = kFont_Regular(12 * kScreenAdaptationRadio);
    [self.cancelButton addTarget:self action:@selector(clickCancel) forControlEvents:UIControlEventTouchUpInside];
    [self.headerView addSubview:self.cancelButton];
    
    self.confirmButton = [UIButton buttonWithType:UIButtonTypeCustom];
    self.confirmButton.frame = CGRectMake(kScreenWidth - 60 * kScreenAdaptationRadio, 0, 60 * kScreenAdaptationRadio, self.headerView.height);
    [self.confirmButton setTitle:@"OK" forState:UIControlStateNormal];
    [self.confirmButton setTitleColor:[UIColor grayColor] forState:UIControlStateNormal];
    self.confirmButton.titleLabel.font = kFont_Regular(12 * kScreenAdaptationRadio);
    [self.confirmButton addTarget:self action:@selector(clickConfirm) forControlEvents:UIControlEventTouchUpInside];
    [self.headerView addSubview:self.confirmButton];
    
    // PickerView
    CGFloat pickViewHeight = 0;
    if (kIs_iPhoneXAndLater) {
        pickViewHeight = pullListHeight - self.headerView.height - 34;
    } else {
        pickViewHeight = pullListHeight - self.headerView.height;
    }
    self.pickerView = [[UIPickerView alloc] initWithFrame:CGRectMake(0, self.headerView.bottom, kScreenWidth, pickViewHeight)];
    self.pickerView.backgroundColor = [UIColor whiteColor];
    self.pickerView.delegate = self;
    self.pickerView.dataSource = self;
    [self.pullListBackgroundView addSubview:self.pickerView];
    
    // Jump to seleted row.
    NSString *selectedName = [recordCurrerntSelectedDic valueForKey:@"showName"];
    NSInteger selectedRow = 0;
    for (int i = 0; i < pullListDataArray.count; i++) {
        NSDictionary *showDic = pullListDataArray[i];
        NSString *showName = [showDic valueForKey:@"showName"];
        if ([selectedName isEqualToString:showName]) {
            selectedRow = i;
            break;
        }
    }
    
    [self.pickerView selectRow:selectedRow inComponent:0 animated:YES];
    
}

/// Click maskView.
- (void)maskClick
{
    [self removeFromSuperview];
    
    [self dismiss];
}

/// Click cancel.
- (void)clickCancel
{
    [self dismiss];
}

/// Click confirm.
- (void)clickConfirm
{
    if (self.completion) {
        self.completion(recordCurrerntSelectedDic);
    }
    [self dismiss];
}

//MARK: - UIPickerViewDelegate
- (CGFloat)pickerView:(UIPickerView *)pickerView widthForComponent:(NSInteger)component
{
    return kScreenWidth;
}

- (CGFloat)pickerView:(UIPickerView *)pickerView rowHeightForComponent:(NSInteger)component
{
    return 35 * kScreenAdaptationRadio;
}

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView
{
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component
{
    return pullListDataArray.count;
}

- (UIView *)pickerView:(UIPickerView *)pickerView viewForRow:(NSInteger)row forComponent:(NSInteger)component reusingView:(UIView *)view
{
    NSDictionary *pullListDic = pullListDataArray[row];
    NSString *showName = [pullListDic valueForKey:@"showName"];
    UILabel *contentLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, kScreenWidth, 35 * kScreenAdaptationRadio)];
    contentLabel.text = showName;
    contentLabel.textAlignment = NSTextAlignmentCenter;
    contentLabel.font = kFont_Regular(16 * kScreenAdaptationRadio);
    return contentLabel;
}


- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component
{
    recordCurrerntSelectedDic = pullListDataArray[row];
}


- (void)show
{
    self.pullListBackgroundView.top = kScreenHeight;
    [UIView animateWithDuration:0.2 animations:^{
        self.pullListBackgroundView.top = kScreenHeight - self.pullListBackgroundView.height;
    } completion:^(BOOL finished) {
        
    }];
    
    UIWindow *window = [UIApplication sharedApplication].delegate.window;
    [window addSubview:self];
}

- (void)dismiss
{

    [UIView animateWithDuration:0.2 animations:^{
        self.pullListBackgroundView.top = kScreenHeight;
    } completion:^(BOOL finished) {
        if (finished) {
            [self removeFromSuperview];
        }

    }];
}

@end

File: ios/Swift/GeneralSettingsSwift/GeneralSettingsSwift/ViewControllers/Views/PullListView.swift
//
//  PullListView.swift
//  GeneralSettingsSwift
//
//  Created by dynamsoft's mac on 2022/12/6.
//

import UIKit

typealias SelectCompletion = (_ selectedDic: [String: Any]) -> Void

class PullListView: UIView, UIPickerViewDelegate, UIPickerViewDataSource {

    private var pullListArray: [[String: Any]]  = []
    
    private var title: String = ""
    
    private var selectedDicInfo: [String :Any] = [:]
    
    private var selectCompletion: SelectCompletion?
    
    let pullListHeight: CGFloat = kIs_iPhoneXAndLater ? (270 + 34) : 270
    
    private lazy var backgroundMaskView: UIView = {
        let view = UIView.init(frame: CGRect(x: 0, y: 0, width: self.width, height: self.height))
        view.backgroundColor = .black.withAlphaComponent(0.2)
        view.isUserInteractionEnabled = true
        return view
    }()
    
    private lazy var pullListBackgroundView: UIView = {
        let view = UIView.init(frame: CGRect(x: 0, y: self.height - pullListHeight, width: self.width, height: pullListHeight))
        view.backgroundColor = .white
        return view
    }()
    
    private lazy var headerView: UIView = {
        let view = UIView.init(frame: CGRect(x: 0, y: 0, width: self.width, height: 40))
        view.backgroundColor = .black.withAlphaComponent(0.05)
        return view
    }()
    
    private lazy var titleLabel: UILabel = {
        let label = UILabel.init(frame: CGRect(x: (kScreenWidth - 100) / 2.0, y: 0, width: 100, height: self.headerView.height))
        label.text = self.title
        label.textColor = .gray
        label.font = kFont_Regular(13)
        label.textAlignment = .center
        return label
    }()
    
    private lazy var cancelButton: UIButton = {
        let button = UIButton.init(frame: CGRect(x: 0, y: 0, width: 60, height: self.headerView.height))
        button.setTitle("Cancel", for: .normal)
        button.setTitleColor(.gray, for: .normal)
        button.titleLabel?.font = kFont_Regular(12)
        button.addTarget(self, action: #selector(cancelAction(_:)), for: .touchUpInside)
        return button
    }()
    
    private lazy var confirmButton: UIButton = {
        let button = UIButton.init(frame: CGRect(x: kScreenWidth - 60, y: 0, width: 60, height: self.headerView.height))
        button.setTitle("OK", for: .normal)
        button.setTitleColor(.gray, for: .normal)
        button.titleLabel?.font = kFont_Regular(12)
        button.addTarget(self, action: #selector(confirimAction(_:)), for: .touchUpInside)
        return button
    }()
    
    private lazy var pickerView: UIPickerView = {
        let pickerViewHeight: CGFloat = kIs_iPhoneXAndLater ? (pullListHeight - self.headerView.height - 34) : (pullListHeight - self.headerView.height)
        let picker = UIPickerView.init(frame: CGRect(x: 0, y: self.headerView.bottom, width: kScreenWidth, height: pickerViewHeight))
        picker.backgroundColor = .white
        picker.delegate = self
        picker.dataSource = self
        return picker
    }()
    
    static func showPullList(with pullListArray:[[String: Any]],
                             title: String,
                             selectedDicInfo:[String: Any],
                             completion: @escaping SelectCompletion) -> Void {
        let pullListView = PullListView(frame: UIScreen.main.bounds, pullListArray: pullListArray, title: title, selectedDicInfo: selectedDicInfo, completion: completion)
        pullListView.show()
    }

    init(frame: CGRect,
         pullListArray:[[String: Any]],
         title: String,
         selectedDicInfo:[String: Any],
         completion: @escaping SelectCompletion) {
        super.init(frame: frame)
        self.pullListArray.append(contentsOf: pullListArray)
        self.title = title
        self.selectedDicInfo = selectedDicInfo
        self.selectCompletion = completion
        setupUI()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        fatalError("init(coder:) has not been implemented")
    }
    
    private func setupUI() -> Void {
        self.addSubview(backgroundMaskView)
        self.backgroundMaskView.addSubview(pullListBackgroundView)
        self.pullListBackgroundView.addSubview(headerView)
        self.headerView.addSubview(titleLabel)
        self.headerView.addSubview(cancelButton)
        self.headerView.addSubview(confirmButton)
        self.pullListBackgroundView.addSubview(pickerView)
        
        // Jump to seleted row.
        let selectedName = selectedDicInfo["showName"] as! String
        var selectedRow = 0
        var currentLoopIndex = 0
        for singleInfoDic in pullListArray {
            let singleName = singleInfoDic["showName"] as! String
            if singleName == selectedName {
                selectedRow = currentLoopIndex
                break
            }
            currentLoopIndex = currentLoopIndex + 1
        }
        
        self.pickerView.selectRow(selectedRow, inComponent: 0, animated: true)
    }
    
    private func show() -> Void {
        self.pullListBackgroundView.top = kScreenHeight
        UIView.animate(withDuration: 0.2) {
            self.pullListBackgroundView.top = kScreenHeight - self.pullListBackgroundView.height
        } completion: { finished in
            
        }

        let window = UIApplication.shared.delegate?.window
        window??.addSubview(self)
    }
    
    private func dismiss() -> Void {
        UIView.animate(withDuration: 0.2) {
            self.pullListBackgroundView.top = kScreenHeight
        } completion: { finished in
            self.removeFromSuperview()
        }
    }
    
    @objc private func cancelAction(_ button: UIButton) -> Void {
        self.dismiss()
    }
    
    @objc private func confirimAction(_ button: UIButton) -> Void {
        self.selectCompletion?(self.selectedDicInfo)
        self.dismiss()
    }
    
    //MARK: - UIPickerViewDelegate
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return pullListArray.count
    }
    
    func pickerView(_ pickerView: UIPickerView, widthForComponent component: Int) -> CGFloat {
        return kScreenWidth
    }
    
    func pickerView(_ pickerView: UIPickerView, rowHeightForComponent component: Int) -> CGFloat {
        return 35
    }
    
    func pickerView(_ pickerView: UIPickerView, viewForRow row: Int, forComponent component: Int, reusing view: UIView?) -> UIView {
        let infoDic = pullListArray[row];
        let showName = infoDic["showName"] as! String
        let contentLabel = UILabel.init(frame: CGRect(x: 0, y: 0, width: kScreenWidth, height: 35))
        contentLabel.text = showName
        contentLabel.textAlignment = .center
        contentLabel.font = kFont_Regular(16)
        return contentLabel
    }
    
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        selectedDicInfo = pullListArray[row]
    }
    
    
}
